File: gaeunji/Week1/실패율.py
def solution(N, stages):
    answer = [0]
    
    # 0, 1, ..., N+1 푸하하하 
    stages_count = [0] * (N+2)
    
    # 0, 1, ..., stage의 길이 
    for i in range(len(stages)):
        stages_count[stages[i]] += 1
    # [0, 1, 3, 2, 1, 0, 1]
    
    prev = len(stages)
    for i in range(1, N+1):
        if prev > 0:
            answer.append(stages_count[i] / prev)
            prev -= stages_count[i]
        else:
            answer.append(0)
    
    sorted_indexes = sorted(range(1, N+1), key=lambda i: answer[i], reverse=True)
    
    return sorted_indexes

# [2, 1, 2, 6, 2, 4, 3, 3]

File: Sjaize/Week1/실패율.java
import java.util.stream.IntStream;
import java.util.Comparator;

class Solution {
    public int[] solution(int N, int[] stages) {
        int[] numPerson = new int[N+2];
        
        for (int person = 0; person < stages.length; person++) {
            int stage = stages[person];
            numPerson[stage]++;
        }
        
        double[] rates = new double[N+2];
        int total_person = stages.length;
        int eliminated = 0;
        
        for (int stage = 1; stage < N+1; stage++) {
            rates[stage] = (double) numPerson[stage] / (total_person - eliminated);
            eliminated = eliminated + numPerson[stage];
            if (eliminated == total_person) {
                break;
            }
        }
        
        return IntStream.range(1, N+1)
                        .boxed()
                        .sorted((a, b) -> Double.compare(rates[b], rates[a]))
                        .mapToInt(i -> i)
                        .toArray();
    }
}
